group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java

```diff
@@ -0,0 +1,958 @@
+     */
+    public boolean isGenericGroup() {
+        return protocolType.map(type ->
+            type.equals(ConsumerProtocol.PROTOCOL_TYPE)
+        ).orElse(false);
+    }
+
+    public void add(GenericGroupMember member) {
+        add(member, null);
+    }
+
+    /**
+     * Add a member to this group.
+     *
+     * @param member the member to add.
+     * @param future the future to complete once the join group phase completes.
+     */
+    public void add(GenericGroupMember member, CompletableFuture<JoinGroupResponseData> future) {
+        member.groupInstanceId().ifPresent(instanceId -> {
+            if (staticMembers.containsKey(instanceId)) {
+                throw new IllegalStateException("Static member with groupInstanceId=" +
+                    instanceId + " cannot be added to group " + groupId + " since" +
+                    " it is already a member.");
+            }
+            staticMembers.put(instanceId, member.memberId());
+        });
+
+        if (members.isEmpty()) {
+            this.protocolType = Optional.of(member.protocolType());
+        }
+
+        if (!Objects.equals(this.protocolType.orElse(null), member.protocolType())) {
+            throw new IllegalStateException("The group and member's protocol type must be the same.");
+        }
+
+        if (!supportsProtocols(
+            member.protocolType(),
+            GenericGroupMember.plainProtocolSet(member.supportedProtocols()))) {
+
+            throw new IllegalStateException("None of the member's protocols can be supported.");
+        }
+
+        if (!leaderId.isPresent()) {
+            leaderId = Optional.of(member.memberId());
+        }
+
+        members.put(member.memberId(), member);
+        incrementSupportedProtocols(member);
+        member.setAwaitingJoinFuture(future);
+
+        if (member.isAwaitingJoin()) {
+            numMembersAwaitingJoinResponse++;
+        }
+
+        pendingJoinMembers.remove(member.memberId());
+    }
+
+    /**
+     * Remove a member from the group.
+     *
+     * @param memberId the member id to remove.
+     */
+    public void remove(String memberId) {
+        GenericGroupMember removedMember = members.remove(memberId);
+        if (removedMember != null) {
+            decrementSupportedProtocols(removedMember);
+            if (removedMember.isAwaitingJoin()) {
+                numMembersAwaitingJoinResponse--;
+            }
+
+            removedMember.groupInstanceId().ifPresent(staticMembers::remove);
+        }
+
+        if (isLeader(memberId)) {
+            Iterator<String> iter = members.keySet().iterator();
+            String newLeader = iter.hasNext() ? iter.next() : null;
+            leaderId = Optional.ofNullable(newLeader);
+        }
+
+        pendingJoinMembers.remove(memberId);
+        pendingSyncMembers.remove(memberId);
+    }
+
+    /**
+     * Check whether current leader has rejoined. If not, try to find another joined member to be
```

Comment:
nit: Check whether the* current leader has rejoined. If not, try to find a new leader among the other joined members.